src/shared/Network/AsyncSocket.hpp

```diff
@@ -0,0 +1,140 @@
+/*
+* This file is part of the CMaNGOS Project. See AUTHORS file for Copyright information
+*
+* This program is free software; you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation; either version 2 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program; if not, write to the Free Software
+* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+*/
+
+#ifndef __HIGH_VOLUME_SOCKET_HPP_
+#define __HIGH_VOLUME_SOCKET_HPP_
+
+#include "Platform/Define.h"
+#include <boost/asio.hpp>
+#include <boost/enable_shared_from_this.hpp>
+#include "boost/lexical_cast.hpp"
+#include "Log.h"
+
+namespace MaNGOS
+{
+    // this socket is different in that it blocks on reads
+    template <typename SocketType>
+    class AsyncSocket : public boost::enable_shared_from_this<SocketType>
```

Comment:
Because when I did the async socket and listener 2y back, this is exactly what boost recommends doing in their documentation. Also, because this new one can actually be used without persistence in lets say session (which might be useful in the future if I try to push for the async maps thing)